GenderPayGap.WebUI/Controllers/Account/ChangePasswordNewController.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System.Linq;
+using System.Threading.Tasks;
+using GenderPayGap.Core.Interfaces;
+using GenderPayGap.Database;
+using GenderPayGap.WebUI.BusinessLogic.Abstractions;
+using GenderPayGap.WebUI.Classes;
+using GenderPayGap.WebUI.Helpers;
+using GenderPayGap.WebUI.Models.Account;
+using GenderPayGap.WebUI.Repositories;
+using GenderPayGap.WebUI.Services;
+using GovUkDesignSystem;
+using GovUkDesignSystem.Parsers;
+using Microsoft.AspNetCore.Mvc;
+
+namespace GenderPayGap.WebUI.Controllers.Account
+{
+    [Route("manage-account")]
+    public class ChangePasswordNewController : Controller
+    {
+
+        private readonly IDataRepository dataRepository;
+        private readonly IUserRepository userRepository;
+        private readonly EmailSendingService emailSendingService;
+
+        public ChangePasswordNewController(
+            IDataRepository dataRepository,
+            IUserRepository userRepository,
+            EmailSendingService emailSendingService)
+        {
+            this.dataRepository = dataRepository;
+            this.userRepository = userRepository;
+            this.emailSendingService = emailSendingService;
+        }
+        
+        // The user asks to change their password
+        // We direct them to a page that asks for their existing and new passwords
+        [HttpGet("change-password-new")]
+        public IActionResult ChangePasswordGet()
+        {
+            ControllerHelper.ThrowIfAdminIsImpersonatingUser(User);
+
+            // Return the Change Personal Details form
+            return View("ChangePasswordNew", new ChangePasswordNewViewModel());
+        }
+
+        [HttpPost("change-password-new")]
+        [ValidateAntiForgeryToken]
+        [PreventDuplicatePost]
+        public async Task<IActionResult> ChangePasswordPost(ChangePasswordNewViewModel viewModel)
+        {
+            // Check all values are provided and NewPassword is at least 8 characters long
+            viewModel.ParseAndValidateParameters(Request, m => m.CurrentPassword);
+            viewModel.ParseAndValidateParameters(Request, m => m.NewPassword);
+            viewModel.ParseAndValidateParameters(Request, m => m.ConfirmNewPassword);
+            
+            // Get the current user
+            User currentUser = ControllerHelper.GetGpgUserFromAspNetUser(User, dataRepository);
+            
+            // Check that passwords are valid
+            ValidatePasswords(viewModel, currentUser);
+
+            if (viewModel.HasAnyErrors())
+            {
+                return View("ChangePasswordNew", viewModel);
+            }
+            
+            userRepository.UpdateUserPasswordUsingPBKDF2(currentUser, viewModel.NewPassword);

```

Comment:
Let's call
```csharp
userRepository.UpdatePassword(currentUser, viewModel.NewPassword);
```
instead of
```csharp
userRepository.UpdateUserPasswordUsingPBKDF2(currentUser, viewModel.NewPassword);
```
because it also audit logs the change

Reply:
Is there a reason `UpdateUserPasswordUsingPBKDF2` is public since we're likely to always want to audit log the change? I picked the wrong one here because I didn't realise that one was called within the other!

Reply:
Good point - no, I think we could:
* change the only usage of it from outside the class to use `UpdatePassword`
* remove `UpdateUserPasswordUsingPBKDF2` from the interface
* make `UpdateUserPasswordUsingPBKDF2` private